Added external_custom_target

SAY-droid427 opened the pull request:
The following tasks have been performed:
- external_custom_target folder added to external directory
- function added to the common_functions to read the test.properties file

smlambert commented on external/external_custom/dockerfile/test.sh:
```diff
@@ -0,0 +1,53 @@
+
+# export MAVEN_OPTS="-Xmx1g"
+
+cd /flink
```
this is specific to repo that was cloned, so should not hardcode to flink

SAY-droid427 replied:
We should use another variable for this and for getting the value for this can we set up a variable DIR in the test.properties file and read from it?

smlambert replied:
this will always be the name of the external test repo that we cloned, so we can figure it out based on the value of that variable (which I think you are currently calling EXTERNAL_CUSTOM_TARGET but really that is a repo_url... probably be clearer to call that EXTERNAL_CUSTOM_REPO

so, for https://github.com/apache/camel-quarkus.git, cd /[camel-quarkus](https://github.com/adoptium/aqa-tests/blob/master/external/camel/dockerfile/camel-test.sh#L45)  and so forth

llxia replied:
I think we can just reuse the cmd to get folder name from git URL
https://github.com/adoptium/aqa-tests/blob/14f5a8348e22e5616dea2ea001ab8d693cddfeee/external/dockerfile_functions.sh#L450

SAY-droid427 replied:
@smlambert Yes, I will change the name of the variable @lxia Yes we should get the command from the git URL. I will make the changes. Thanks!

smlambert commented on external/external_custom/dockerfile/test.sh:
```diff
@@ -0,0 +1,53 @@
+cd /flink
+pwd
+echo "Compile and run external_custom tests"
+mvn clean install
```
This is the EXTERNAL_TEST_CMD variable which should be passed in for cases where its not mvn clean install (which we can use as the default command, as its used most often by applications testing).

SAY-droid427 replied:
For this  we have to add the properties in the test.properties of the individual tests and get the values from the test.properties of the test being run and use it in the common_functions.sh, according to my understanding. Is this correct?

smlambert replied:
yes, that is what I was thinking we do, for tests in category 1 (unsuitable for external_custom), you do not define test_command in test.properties (because there are many commands), for those, we need to keep their Dockerfile/xx-test.sh file which contains all the special set up, for tests in category 2 (using mvn clean install) or category 3 (some other simple one line command) those can be added to test_command in test.properties file

for the contents of external_custom/test.properties, then test_command="${EXT_TEST_CMD}" if EXT_TEST_CMD is defined, if not, we use the default value 'mvn clean install'

smlambert commented on external/external_custom/dockerfile/test.sh:
```diff
@@ -0,0 +1,53 @@
+test_exit_code=$?
+echo "Build external_custom completed"
+
+find ./ -type d -name 'surefire-reports' -exec cp -r "{}" /testResults \;
```
we likely can not assume everything will go into 'surefire-reports' dir, but we can address this in a later PR.

smlambert commented on external/common_functions.sh:
```diff
@@ -219,6 +226,23 @@ function set_test_info() {
         ubi_packages="git wget unzip"
         ubi_minimal_packages="${ubi_packages}"
         ;;
+    external_custom)
```
Does it complicate things to put this content into a test.properties file in the external_custom dir and read from there?  for now we can leave as is, but ultimately the set_test_info() function would be removed completely, and we would read_test_info from each test.properties file found test subdir.

SAY-droid427 replied:
I will add a test.properties file in the external_custom dir and add all the contents there for now. After all the variables have been set up and working, the set_test_info() would be removed. I think this should work.

smlambert commented on external/common_functions.sh:
```diff
@@ -219,6 +226,23 @@ function set_test_info() {
+        github_url="${EXTERNAL_CUSTOM_TARGET}"
+        script="test.sh"
+        test_results="testResults"
+        tag_version="master"
```
tag_version="${EXT_REPO_BRANCH}"

SAY-droid427 replied:
I have made the required changes for the EXT_REPO_BRANCH and successfully run the spring tests too which couldn't be run earlier due to branch name issues.
![spring](https://user-images.githubusercontent.com/59281636/123113206-fd32a800-d45b-11eb-9af6-c2d5a49237ea.png)

smlambert commented on external/common_functions.sh:
```diff
@@ -219,6 +226,23 @@ function set_test_info() {
+        PROPERTY_FILE=external_custom/test.properties
+        GITHUB_URL=$(getProperty "github_url")
+        github_url="${EXTERNAL_CUSTOM_TARGET}"
+        script="test.sh"
```
test_command="${EXTERNAL_TEST_CMD}"

smlambert reviewed (commented):
Some notes about how to handle branch name and test command.

SAY-droid427 commented:
@smlambert I have added the test_command property to the  test.properties file of eligible tests. I have also changed the name of the EXTERNAL_CUSTOM_TARGET to EXTERNAL_CUSTOM_REPO.
As discussed in the meet I have to add an external directory where the repo has to be cloned and then we can cd into that EXTERNAL_DIR(or any name) directory. But I couldn't get in which file I have to make the change. If you could please point that out a bit. Thanks!

SAY-droid427 commented:
@smlambert @llxia I have made all the changes as discussed in the meeting. Please review the changes and let me know of any other changes needed.

llxia commented on external/dockerfile_functions.sh:
```diff
@@ -451,7 +451,7 @@ print_clone_project() {
     echo -e "# Clone ${test} source" \
             "\nENV ${test_tag}=\$${test_tag}" \
-            "\nRUN git clone ${github_url}" \
+            "\nRUN git clone ${github_url} external_test_repo" \
```
I believe this is a common script. Changing the folder name here means we change the behavior for all external tests - all external tests check out into `external_test_repo` folder. Therefore, all external tests need to cd into `external_test_repo` in their shell script, not just external_custom test. 

On the second thought, it would be easier to figure out the repo dir for external_custom. Can you get EXTERNAL_CUSTOM_REPO in test.sh? If so, just use cmd to figure out the folder (similar to 450).

SAY-droid427 replied:
Since we have set the EXTERNAL_CUSTOM_TARGET as an environment variable, I think we can get the EXTERNAL_CUSTOM_REPO from the EXTERNAL_CUSTOM_TARGET using a function in test.sh itself.

SAY-droid427 replied:
@lxia I have changed the logic of the external_test_repo as you told but one thing I noticed is that while moving into the jacoco folder, we have to cd into the following:[ https://github.com/adoptium/aqa-tests/blob/b766eb2c81252e8beefa3dcdf7e8c99c99210b68/external/jacoco/dockerfile/jacoco-test.sh#L41](https://github.com/adoptium/aqa-tests/blob/b766eb2c81252e8beefa3dcdf7e8c99c99210b68/external/jacoco/dockerfile/jacoco-test.sh#L41). So is it necessary to cd into the org.jacoco.build subfolder?

smlambert replied:
interesting, so jacoco might be an exception case (not as suitable to use this external_custom approach as others, because it doesn't conform to the ideal scenario... where ideal is clone repository, cd into repo dir, run a single command to execute tests)

you can check if its necessary by seeing what happens if you cd into repo name (and not subfolder), are you still able to run tests or does it fail to work.

SAY-droid427 replied:
@smlambert So I run the tests and it appears that we don't really need to cd into the org.jacoco.build folder again as the tests run successfully.
![sfd](https://user-images.githubusercontent.com/59281636/123460428-3d815a00-d605-11eb-9c78-231ff0d92aba.png)

smlambert replied:
very good, thanks for confirming!

SAY-droid427 replied:
Sorry for the misinformation i provided earlier! Although the make compile target can be run successfully, but the make _jacoco_test cannot be run without cd ing into the org.jacoco folder. So I think this cannot be included in the external_custom target tests.

SAY-droid427 commented:
@smlambert @lxia Could you please let me know what should I do next now that the changes have been made? Thanks!

smlambert reviewed (approved):
Looks like we are at the point where we can and should merge this.  Then we can use the other branch from https://github.com/adoptium/aqa-tests/pull/2687 to test with in Grinder_Tweetest job.  Thanks @SAY-droid427 !

smlambert commented:
@llxia - anything else you'd like to see here before we merge?  (we expect other incremental improvements to follow also, but merging this enables Grinder_Tweetest testing of https://github.com/adoptium/aqa-tests/pull/2687.

llxia commented on external/external_custom/dockerfile/test.sh:
```diff
@@ -0,0 +1,50 @@
+
+external_test_repo="$(echo ${EXTERNAL_CUSTOM_REPO} | awk -F'/' '{print $NF}' | sed 's/.git//g')"
+
+cd /external_test_repo
```
Does this work without `$`?

SAY-droid427 replied:
Yes the `make compile` works fine but `make _external_custom` does not work both with and without $. The whole output:
![output1](https://user-images.githubusercontent.com/59281636/124610131-daf24e80-de8d-11eb-9304-bfbb688f8acd.png)
![output2](https://user-images.githubusercontent.com/59281636/124610163-e2b1f300-de8d-11eb-9b45-c839394ac37c.png)
![output3](https://user-images.githubusercontent.com/59281636/124610192-ea719780-de8d-11eb-9ba1-f87b2a0591fb.png)
![output4](https://user-images.githubusercontent.com/59281636/124610220-f0677880-de8d-11eb-9eac-75d1d2918677.png)
![output7](https://user-images.githubusercontent.com/59281636/124610247-f6f5f000-de8d-11eb-878b-33520e43d2d3.png)
![output8](https://user-images.githubusercontent.com/59281636/124610278-ff4e2b00-de8d-11eb-9a75-44b984a08266.png)
![output9](https://user-images.githubusercontent.com/59281636/124610320-083efc80-de8e-11eb-97c7-b7f372b16824.png)

llxia replied:
The `test.sh` will run in the actual test execution (make _external_custom), not `make compile`. 
I will approve this PR, so we can run this in Grinder. Instead of debugging in your local, I think it will be easier to debug in Grinder.

SAY-droid427 replied:
Thanks @llxia ! Could you please guide me a bit on how to run this in Grinder?

llxia replied:
By Grinder, I mean https://ci.adoptopenjdk.net/job/Grinder_Tweetest/. Once #2687 is merged, we can run tests on Jenkins.
